sdk/src/metrics/multi_observer_result.cc

```diff
@@ -0,0 +1,154 @@
+#include "opentelemetry/nostd/variant.h"
+#include "opentelemetry/sdk/common/global_log_handler.h"
+#include "opentelemetry/sdk/metrics/async_instruments.h"
+#include "opentelemetry/sdk/metrics/multi_observer_result.h"
+#include "opentelemetry/sdk/metrics/observer_result.h"
+#include "opentelemetry/sdk/metrics/state/metric_storage.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace metrics
+{
+
+namespace
+{
+struct StoreResultVisitor
+{
+  void operator()(ObserverResultT<int64_t> &result) const
+  {
+    storage->RecordLong(result.GetMeasurements(), collection_ts);
+  }
+
+  void operator()(ObserverResultT<double> &result) const
+  {
+    storage->RecordDouble(result.GetMeasurements(), collection_ts);
+  }
+
+  void operator()(const nostd::monostate &) const {}
+
+  AsyncWritableMetricStorage *storage;
+  opentelemetry::common::SystemTimestamp collection_ts;
+};
+}  // namespace
+
+void MultiObserverResult::RegisterInstrument(
+    opentelemetry::metrics::ObservableInstrument *instrument)
+{
+  observer_results_.emplace(instrument, nostd::monostate());
+}
+
+void MultiObserverResult::DeregisterInstrument(
+    opentelemetry::metrics::ObservableInstrument *instrument)
+{
+  observer_results_.erase(instrument);
+}
+
+size_t MultiObserverResult::InstrumentCount() const
+{
+  return observer_results_.size();
+}
+
+bool MultiObserverResult::HasInstrument(
+    const opentelemetry::metrics::ObservableInstrument *instrument) const
+{
+  return observer_results_.find(const_cast<opentelemetry::metrics::ObservableInstrument *>(
+             instrument)) != observer_results_.end();
+}
+
+void MultiObserverResult::GetInstruments(
+    nostd::function_ref<void(opentelemetry::metrics::ObservableInstrument *)> callback)
+{
+  for (auto &el : observer_results_)
+  {
+    callback(el.first);
+  }
+}
+
+void MultiObserverResult::Reset()
+{
+  for (auto it = observer_results_.begin(); it != observer_results_.end(); ++it)
+  {
+    it->second = nostd::monostate();
+  }
+}
+
+void MultiObserverResult::StoreResults(opentelemetry::common::SystemTimestamp collection_ts)
+{
+  for (auto &el : observer_results_)
+  {
+    auto *instrument = el.first;
+    auto &result     = el.second;
+
+    auto storage = static_cast<opentelemetry::sdk::metrics::ObservableInstrument *>(instrument)
+                       ->GetMetricStorage();
+    nostd::visit(StoreResultVisitor{storage, collection_ts}, result);
+  }
+}
+
+opentelemetry::metrics::ObserverResultT<double> &MultiObserverResult::ForInstrumentDouble(
+    const opentelemetry::metrics::ObservableInstrument *instrument)
+{
+  static opentelemetry::sdk::metrics::ObserverResultT<double> null_result;
+  // const_cast is appropriate here, because we're _not_ modifying the passed-in pointer;
+  // we just need to make it non-const to be able to look it up in our map.
+  auto it = observer_results_.find(
+      const_cast<opentelemetry::metrics::ObservableInstrument *>(instrument));
```

Comment:
Not sure if it is feasible, but could the `observer_results_` map use `const opentelemetry::metrics::ObservableInstrument *` instead, and register / unregister use const instruments as well ?

Reply:
No need to change all the code, I just would like to discuss if using const instruments is a) possible and b) desirable.